### ISSUE TITLE
"No template found" error occurring 'randomly' in system containing multiple patches after 'hacking' _generatePatchedSingleResidueTemplates

### ISSUE THREAD

**Issue by jchus** [openmm/openmm#3172]
I have a system with several patches (5' and 3' end of RNA, modified RNA bases, N and C terminals of protein, Zn-bound residues,...).
In order for system = forcefield.createSystem(self.topology) to succeed, I had to change the function _generatePatchedSingleResidueTemplates in forcefield.py and use:
if (index+1 < min([len(patches), 12])):
instead of 
if index+1 < len(patches):

otherwise it would never complete and the memory usage of the process would keep increasing (because of the recursivity I suppose). '12' seemed to be the right balance for all my patches to be applied without taking too much time.
With that modification, my system is loaded correctly, 9 times out of 10. Every so often, I am getting a "No template found" error that occurs on a patched residue.
Would you have any advice to troubleshoot this? This is hard to find the issue since this happen infrequently and I have only a superficial understanding of the code.

**Comment by peastman** [comment]
In principle the number of templates generated by `_generatePatchedSingleResidueTemplates()` can be exponential in the number of patches that can be applied to a single residue, though in practice it's usually less than that.  And most residues only support a small number of patches, though there are exceptions.

Can you post the files needed to reproduce this?

**Comment by jchus** [comment]
I can't post the files here but I can share them privately.

**Comment by peastman** [comment]
That would be great, thanks.

**Comment by peastman** [comment]
Where did you get your force field XML file from?  It lists many more patches than the standard one.  Consider the FLUB residue.  The standard charm36.xml file from OpenMM only has two `<AllowPatch>` tags for it, MET1 and MET2.  But your file lists 73 different patches for it, from which `_generatePatchedSingleResidueTemplates()` generates 96,263 templates.

**Comment by jchus** [comment]
That's our in-house forcefield based on charmm36. We do have quite a few extra patches. Shouldn't it read only the patches for the residues contained in the PDB file? Or do I need to create a custom forcefield each time by pruning our general forcefield to the current system?

**Comment by peastman** [comment]
It doesn't know in advance what patches will be needed to match the residues in the PDB file.  It first tries to match all the standard templates.  If a residue doesn't have a match, it starts applying patches to templates in the hope of creating a match.  In the case above, there's one template that can take 73 different patches.  Is it possible some combination of those patches will lead to a match?  It doesn't know until it tries.

**Comment by jchus** [comment]
Yes, of course. I guess dumping our full forcefield into xml wasn't the best idea. We can easily change that and have a forcefield that only contains residues and patches that show up in our pdb file. I am still not sure why it fails 1 out of 10 times in our case but that's an aside. Thank you for looking at this issue and pointing out the problem.

**Comment by peastman** [comment]
There's probably some nondeterministic order going on somewhere.  When you limit it to only the first 12 patches, it won't necessarily be the same 12 every time unless you can guarantee they'll always be enumerated in the same order.
